monads/src/FailCont/Main.hs

```diff
@@ -1,26 +1,67 @@
 module FailCont.Main where
 
-import FailCont.FailCont 
+import FailCont.FailCont
 import Text.Read (readMaybe)
 
-data Error 
-  = EmptyInput 
-  | ParseFailed String 
+data Error
+  = EmptyInput
+  | ParseFailed String
   | DivisionByZero
   deriving (Show, Eq)
 
+parser :: String -> Either Error Int
+parser s
+  | null s = Left EmptyInput
+  | otherwise = maybe (Left (ParseFailed s)) Right (readMaybe s)
+
 addInts :: String -> String -> FailCont r Error Int
-addInts = undefined 
+addInts s1 s2 =
+  toFailCont
+    ( do
+        a <- parser s1
+        b <- parser s2
+        Right (a + b)
+    )
+
+divInts :: String -> String -> FailCont r Error Int
+divInts s1 s2 =
+  toFailCont
+    ( do
+        a <- parser s1
+        b <- parser s2
+        if b == 0 then Left DivisionByZero else Right (a `div` b)
+    )
+
+sqrtInts :: String -> FailCont r Error Double
+sqrtInts s =
+  toFailCont
+    ( do
+        n <- parser s
+        if n < 0 then Left (ParseFailed s) else Right (sqrt (fromIntegral n))
```

Comment:
Why do you report a parsing error when the number is negative?

Reply:
it slipped my mind to create another error for negative numbers and I just threw a parsing error - fixed it